File: tsp_algorithms.py
"""
TSP Algorithm Implementations
Contains Nearest Neighbor, Genetic Algorithm, and Dynamic Programming solutions
"""

import random
import math
from typing import List, Tuple, Dict

class TSPSolver:
    """Solver for Traveling Salesman Problem"""
    
    def __init__(self, cities, edges):
        """
        Initialize solver
        cities: List of City objects
        edges: Dict of (city1, city2): weight
        """
        self.cities = cities
        self.edges = edges
    
    def get_distance(self, city1, city2):
        """Get distance between two cities"""
        return self.edges.get((city1, city2), float('inf'))
    
    def calculate_path_cost(self, path):
        """Calculate total cost of a path"""
        cost = 0
        for i in range(len(path)):
            city1 = path[i]
            city2 = path[(i + 1) % len(path)]
            cost += self.get_distance(city1, city2)
        return cost
    
    def nearest_neighbor(self, start_city=None):
        """
        Nearest Neighbor Algorithm (Greedy Approach)
        Time Complexity: O(n^2)
        """
        if not self.cities:
            return [], 0
        
        if start_city is None:
            start_city = self.cities[0]
        
        unvisited = set(self.cities)
        path = [start_city]
        current = start_city
        unvisited.remove(current)
        
        while unvisited:
            nearest = None
            min_dist = float('inf')
            
            for city in unvisited:
                dist = self.get_distance(current, city)
                if dist < min_dist:
                    min_dist = dist
                    nearest = city
            
            if nearest is None:
                break
            
            path.append(nearest)
            current = nearest
            unvisited.remove(current)
        
        total_cost = self.calculate_path_cost(path)
        return path, total_cost
    
    def genetic_algorithm(self, population_size=100, generations=500, mutation_rate=0.01):
        """
        Genetic Algorithm for TSP
        Uses selection, crossover, and mutation
        """
        if len(self.cities) < 2:
            return [], 0
        
        def create_individual():
            """Create a random tour"""
            tour = self.cities.copy()
            random.shuffle(tour)
            return tour
        
        def fitness(tour):
            """Fitness is inverse of tour cost"""
            cost = self.calculate_path_cost(tour)
            return 1 / (cost + 1)
        
        def select_parent(population, fitnesses):
            """Tournament selection"""
            tournament_size = 5
            tournament = random.sample(list(zip(population, fitnesses)), 
                                     min(tournament_size, len(population)))
            return max(tournament, key=lambda x: x[1])[0]
        
        def crossover(parent1, parent2):
            """Order crossover (OX)"""
            size = len(parent1)
            start, end = sorted(random.sample(range(size), 2))
            
            child = [None] * size
            child[start:end] = parent1[start:end]
            
            pointer = end
            for city in parent2[end:] + parent2[:end]:
                if city not in child:
                    if pointer >= size:
                        pointer = 0
                    child[pointer] = city
                    pointer += 1
            
            return child
        
        def mutate(tour):
            """Swap mutation"""
            if random.random() < mutation_rate:
                i, j = random.sample(range(len(tour)), 2)
                tour[i], tour[j] = tour[j], tour[i]
            return tour
        
        # Initialize population
        population = [create_individual() for _ in range(population_size)]
        
        best_tour = None
        best_fitness = 0
        
        for generation in range(generations):
            # Calculate fitness
            fitnesses = [fitness(tour) for tour in population]
            
            # Track best
            max_fitness_idx = fitnesses.index(max(fitnesses))
            if fitnesses[max_fitness_idx] > best_fitness:
                best_fitness = fitnesses[max_fitness_idx]
                best_tour = population[max_fitness_idx].copy()
            
            # Create new population
            new_population = []
            
            # Elitism - keep best individual
            new_population.append(best_tour.copy())
            
            while len(new_population) < population_size:
                parent1 = select_parent(population, fitnesses)
                parent2 = select_parent(population, fitnesses)
                
                child = crossover(parent1, parent2)
                child = mutate(child)
                
                new_population.append(child)
            
            population = new_population
        
        total_cost = self.calculate_path_cost(best_tour)
        return best_tour, total_cost
    
    def dynamic_programming(self):
        """
        Dynamic Programming solution using Held-Karp algorithm
        Time Complexity: O(n^2 * 2^n)
        Space Complexity: O(n * 2^n)
        Only practical for small numbers of cities (< 20)
        """
        if len(self.cities) < 2:
            return [], 0
        
        n = len(self.cities)
        
        # For large n, fall back to nearest neighbor
        if n > 20:
            return self.nearest_neighbor()
        
        # Create city index mapping
        city_to_idx = {city: i for i, city in enumerate(self.cities)}
        
        # dp[mask][i] = minimum cost to visit cities in mask ending at city i
        dp = [[float('inf')] * n for _ in range(1 << n)]
        parent = [[None] * n for _ in range(1 << n)]
        
        # Start from city 0
        dp[1][0] = 0
        
        # Iterate through all subsets
        for mask in range(1, 1 << n):
            for last in range(n):
                if not (mask & (1 << last)):
                    continue
                
                if dp[mask][last] == float('inf'):
                    continue
                
                # Try to extend to next city
                for next_city in range(n):
                    if mask & (1 << next_city):
                        continue
                    
                    next_mask = mask | (1 << next_city)
                    dist = self.get_distance(self.cities[last], self.cities[next_city])
                    new_cost = dp[mask][last] + dist
                    
                    if new_cost < dp[next_mask][next_city]:
                        dp[next_mask][next_city] = new_cost
                        parent[next_mask][next_city] = last
        
        # Find best ending city
        full_mask = (1 << n) - 1
        best_cost = float('inf')
        best_last = -1
        
        for last in range(n):
            # Add cost to return to start
            cost = dp[full_mask][last] + self.get_distance(self.cities[last], self.cities[0])
            if cost < best_cost:
                best_cost = cost
                best_last = last
        
        # Reconstruct path
        path = []
        if best_last != -1:
            mask = full_mask
            current = best_last
            
            while current != 0:
                path.append(self.cities[current])
                prev_mask = mask ^ (1 << current)
                current = parent[mask][current]
                mask = prev_mask
            
            path.append(self.cities[0])
            path.reverse()

        return path, best_cost
    
    def two_opt(self, tour):
        """
        2-opt improvement heuristic
        Can be applied to improve any tour
        """
        improved = True
        best_tour = tour.copy()
        best_cost = self.calculate_path_cost(best_tour)
        
        while improved:
            improved = False
            
            for i in range(1, len(tour) - 2):
                for j in range(i + 1, len(tour)):
                    if j - i == 1:
                        continue
                    
                    new_tour = best_tour.copy()
                    new_tour[i:j] = reversed(new_tour[i:j])
                    
                    new_cost = self.calculate_path_cost(new_tour)
                    
                    if new_cost < best_cost:
                        best_tour = new_tour
                        best_cost = new_cost
                        improved = True
                        break
                
                if improved:
                    break
        
        return best_tour, best_cost

File: tsp_algorithms.c
/*
 * TSP Algorithm Implementations
 * Contains Nearest Neighbor, Genetic Algorithm, and Dynamic Programming solutions
 */

#include "tsp_algorithms.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <limits.h>

#define POPULATION_SIZE 100
#define GENERATIONS 500
#define MUTATION_RATE 0.01
#define TOURNAMENT_SIZE 5

static double get_distance(TSPData *data, int i, int j) {
    if (data->edge_exists[i][j]) {
        return data->edges[i][j];
    }
    return INFINITY;
}

static double calculate_path_cost(TSPData *data, int *path, int length) {
    double cost = 0;
    for (int i = 0; i < length; i++) {
        int next = (i + 1) % length;
        cost += get_distance(data, path[i], path[next]);
    }
    return cost;
}

// Nearest Neighbor Algorithm - O(n^2)
void tsp_nearest_neighbor(TSPData *data) {
    if (data->city_count == 0) return;
    
    int visited[MAX_CITIES] = {0};
    int path[MAX_CITIES];
    int path_len = 0;
    
    // Start from city 0
    int current = 0;
    path[path_len++] = current;
    visited[current] = 1;
    
    while (path_len < data->city_count) {
        int nearest = -1;
        double min_dist = INFINITY;
        
        for (int i = 0; i < data->city_count; i++) {
            if (!visited[i]) {
                double dist = get_distance(data, current, i);
                if (dist < min_dist) {
                    min_dist = dist;
                    nearest = i;
                }
            }
        }
        
        if (nearest == -1) break;
        
        path[path_len++] = nearest;
        visited[nearest] = 1;
        current = nearest;
    }
    
    // Store solution
    data->solution_length = path_len;
    for (int i = 0; i < path_len; i++) {
        data->solution_path[i] = &data->cities[path[i]];
    }
    data->total_cost = calculate_path_cost(data, path, path_len);
}

// Genetic Algorithm Implementation
typedef struct {
    int tour[MAX_CITIES];
    double fitness;
} Individual;

static void shuffle_tour(int *tour, int n) {
    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int temp = tour[i];
        tour[i] = tour[j];
        tour[j] = temp;
    }
}

static double fitness_function(TSPData *data, int *tour, int n) {
    double cost = calculate_path_cost(data, tour, n);
    return 1.0 / (cost + 1.0);
}

static void select_parent(Individual *population, int pop_size, int *selected) {
    int best_idx = rand() % pop_size;
    double best_fitness = population[best_idx].fitness;
    
    for (int i = 1; i < TOURNAMENT_SIZE && i < pop_size; i++) {
        int idx = rand() % pop_size;
        if (population[idx].fitness > best_fitness) {
            best_fitness = population[idx].fitness;
            best_idx = idx;
        }
    }
    
    memcpy(selected, population[best_idx].tour, MAX_CITIES * sizeof(int));
}

static void crossover(int *parent1, int *parent2, int *child, int n) {
    int start = rand() % n;
    int end = start + 1 + rand() % (n - start);
    
    int used[MAX_CITIES] = {0};
    
    // Copy segment from parent1
    for (int i = start; i < end; i++) {
        child[i] = parent1[i];
        used[parent1[i]] = 1;
    }
    
    // Fill remaining from parent2
    int pos = end % n;
    for (int i = 0; i < n; i++) {
        int idx = (end + i) % n;
        if (!used[parent2[idx]]) {
            child[pos] = parent2[idx];
            used[parent2[idx]] = 1;
            pos = (pos + 1) % n;
        }
    }
}

static void mutate(int *tour, int n) {
    if ((double)rand() / RAND_MAX < MUTATION_RATE) {
        int i = rand() % n;
        int j = rand() % n;
        int temp = tour[i];
        tour[i] = tour[j];
        tour[j] = temp;
    }
}

void tsp_genetic_algorithm(TSPData *data) {
    if (data->city_count < 2) return;
    
    srand(time(NULL));
    int n = data->city_count;
    
    // Initialize population
    Individual population[POPULATION_SIZE];
    Individual new_population[POPULATION_SIZE];
    
    for (int i = 0; i < POPULATION_SIZE; i++) {
        for (int j = 0; j < n; j++) {
            population[i].tour[j] = j;
        }
        shuffle_tour(population[i].tour, n);
        population[i].fitness = fitness_function(data, population[i].tour, n);
    }
    
    // Find initial best
    int best_idx = 0;
    for (int i = 1; i < POPULATION_SIZE; i++) {
        if (population[i].fitness > population[best_idx].fitness) {
            best_idx = i;
        }
    }
    
    Individual best = population[best_idx];
    
    // Evolution
    for (int gen = 0; gen < GENERATIONS; gen++) {
        // Elitism - keep best
        new_population[0] = best;
        
        // Create new population
        for (int i = 1; i < POPULATION_SIZE; i++) {
            int parent1[MAX_CITIES], parent2[MAX_CITIES];
            select_parent(population, POPULATION_SIZE, parent1);
            select_parent(population, POPULATION_SIZE, parent2);
            
            crossover(parent1, parent2, new_population[i].tour, n);
            mutate(new_population[i].tour, n);
            
            new_population[i].fitness = fitness_function(data, 
                                                         new_population[i].tour, n);
        }
        
        // Update population
        memcpy(population, new_population, sizeof(population));
        
        // Track best
        for (int i = 0; i < POPULATION_SIZE; i++) {
            if (population[i].fitness > best.fitness) {
                best = population[i];
            }
        }
    }
    
    // Store solution
    data->solution_length = n;
    for (int i = 0; i < n; i++) {
        data->solution_path[i] = &data->cities[best.tour[i]];
    }
    data->total_cost = calculate_path_cost(data, best.tour, n);
}

// Dynamic Programming - Held-Karp Algorithm O(n^2 * 2^n)
void tsp_dynamic_programming(TSPData *data) {
    if (data->city_count < 2) return;
    
    int n = data->city_count;
    
    // For large n, fall back to nearest neighbor
    if (n > 20) {
        tsp_nearest_neighbor(data);
        return;
    }
    
    int max_mask = 1 << n;
    
    // Allocate DP table
    double **dp = malloc(max_mask * sizeof(double *));
    int **parent = malloc(max_mask * sizeof(int *));
    
    for (int i = 0; i < max_mask; i++) {
        dp[i] = malloc(n * sizeof(double));
        parent[i] = malloc(n * sizeof(int));
        for (int j = 0; j < n; j++) {
            dp[i][j] = INFINITY;
            parent[i][j] = -1;
        }
    }
    
    // Base case - start from city 0
    dp[1][0] = 0;
    
    // Fill DP table
    for (int mask = 1; mask < max_mask; mask++) {
        for (int last = 0; last < n; last++) {
            if (!(mask & (1 << last))) continue;
            if (dp[mask][last] == INFINITY) continue;
            
            // Try extending to next city
            for (int next = 0; next < n; next++) {
                if (mask & (1 << next)) continue;
                
                int next_mask = mask | (1 << next);
                double dist = get_distance(data, last, next);
                double new_cost = dp[mask][last] + dist;
                
                if (new_cost < dp[next_mask][next]) {
                    dp[next_mask][next] = new_cost;
                    parent[next_mask][next] = last;
                }
            }
        }
    }
    
    // Find best ending city
    int full_mask = (1 << n) - 1;
    double best_cost = INFINITY;
    int best_last = -1;
    
    for (int last = 0; last < n; last++) {
        double cost = dp[full_mask][last] + get_distance(data, last, 0);
        if (cost < best_cost) {
            best_cost = cost;
            best_last = last;
        }
    }
    
    // Reconstruct path
    if (best_last != -1) {
        int path[MAX_CITIES];
        int path_len = 0;
        int mask = full_mask;
        int current = best_last;
        
        while (current != 0) {
            path[path_len++] = current;
            int prev_mask = mask ^ (1 << current);
            current = parent[mask][current];
            mask = prev_mask;
        }
        path[path_len++] = 0;
        
        // Reverse path
        data->solution_length = path_len;
        for (int i = 0; i < path_len; i++) {
            data->solution_path[i] = &data->cities[path[path_len - 1 - i]];
        }
        data->total_cost = best_cost;
    }
    
    // Free memory
    for (int i = 0; i < max_mask; i++) {
        free(dp[i]);
        free(parent[i]);
    }
    free(dp);
    free(parent);
}